### PR TITLE
Spike subtraction instead of interpolation

### PR DESCRIPTION
**Pull request by SeiyaNozaki**
continuation of https://github.com/cta-observatory/cta-lstchain/pull/574 

This PR is to subtract corresponding ADC counts on spike cells instead of interpolation.
(related this issue: https://github.com/cta-observatory/cta-lstchain/issues/571)

In this PR, I added a function to find spike cells, put corresponding flags, and subtract spike ADC counts. 
Please find the attached summary of my checks using this new function.
[20210403_SpikeCorrection.pdf](https://github.com/cta-observatory/ctapipe_io_lst/files/6252797/20210403_SpikeCorrection.pdf)

For now, spike ADC counts are hard-coded, and common values are used in the camera.
But, it might be better to find spike heights for each pixel during the computation of drs pedestals.
(I'd like to hear your comments again @TomohikoOka @tysaito2008 @pawel21 @jsitarek )
I will make another PR in lstchain for that after https://github.com/cta-observatory/cta-lstchain/pull/601

### CODE REVIEW

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:203]: maybe make here a switch, if interpolation or subtraction should be used

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:616]: those hard coded values are indeed not very clear, why not implement reading them from the pedestal file immediatelly? 

also, do you have enough statistics in the general pedestal file to estimate those values, or is a special run needed?
if a special run is needed, and no night-to-night changes are expected, another possibility would be to pass those values via configuration file (with those values as defaults)

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:801]: you use (spike_pos - current_fc) % N_CAPACITORS_CHANNEL in 3 lines, maybe call it a new variable

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:841]: same comment as above, you can introduce a new variable here since it is used 3 times

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:166]: this is only used to tag the changed slices in the code, it is not used for actual calculation, right?

have you checked if the subtraction code and the old interpolation code are similar in speed?
removing those spike_flag tags could speed it up a bit more.

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [inline ctapipe_io_lst/calibration.py:836]: I thought that this condition on the last capacitor should be inverted for the data before and after Nov 2019, while it is the same one here, I checked the interpolation function and the condition there is also the same in both cases, @pawel21  and @SeiyaNozaki can you check that it is really intended like this (in both interpolation and subtraction)

**Review by jsitarek — CHANGES_REQUESTED** [review]
I left a few comments, the only one that is possibly critical is the one about the last capacitor condition between the two subtraction functions. Please check that there is no mistake in the code here.
Other comments are just suggestions.

File: ctapipe_io_lst/calibration.py
Comment (maxnoe) [reply ctapipe_io_lst/calibration.py:616]: See https://github.com/cta-observatory/cta-lstchain/pull/725

the standard 20k Events seems enough if we only need spike heights per pixel and not per capacitor.

File: ctapipe_io_lst/calibration.py
Comment (SeiyaNozaki) [reply ctapipe_io_lst/calibration.py:836]: yep, I expected it like you at that time, but it seems the same condition is needed to predict spike position...
As of now, I don't understand why it is not inverted, but apply this condition because it seems actually correct...

**Comment by SeiyaNozaki** [comment]
Hi @maxnoe,
Let me clarify one thing.
I guess you are developing the codes to subtract spike heights using the corresponding table generated with cta-observatory/cta-lstchain#725 (e.g. get_spike_A_positions method) with other code refactorings.
If so, should I stop here for this PR and review your codes later?

**Comment by maxnoe** [comment]
@SeiyaNozaki Yes, I guess it makes sense to wait until at least #105 is merged, since that makes some relatively large refactoring needed for already gain selected data, which in turn also makes it easier to just implement a different spike correction algorithm.

We can discuss about how best to apply it and from which files and so on, but the splitting up of the large spike correction function into several independent ones makes it much easier. We need much fewer functions that way. 

Instead of all combinations of spike correction method, gain selection method, and firmware version we can just combine them.

File: ctapipe_io_lst/calibration.py
Comment (jsitarek) [reply ctapipe_io_lst/calibration.py:836]: thx for checking @SeiyaNozaki 
either way this part of code is only for the old pre Nov-2019 data, and will likely be removed

**Comment by jsitarek** [comment]
Hi @maxnoe 

about the conflict with #105  - what you split into several functions is the dt correction curve, the spike interpolation was already split. 
Nevertheless, since the changes in @105 are larger than in here, I guess it will be indeed simpler to first merge 105 and then update this one (or rather redo it from scratch, because likely updating this branch after the #105 merge will fail badly)

**Comment by maxnoe** [comment]
@jsitarek I also split the spike correction into the loop over all gains / pixels part and the actual application to a single pixel. Next step is to split the position calculation from the correction application.

**Comment by jsitarek** [comment]
ok @maxnoe, I was confused, because the old code also had the actual interpolation separated into a different function, but I see that you indeed added one more in the chain, i.e. LSTR0Corrections::interpolate_spikes ==> interpolate_spikes/interpolate_spikes_gain_selected ==> interpolate_spikes_pixel/ interpolate_spikes_pixel_old_firmware ==>  interpolate_spike_A Let me comment on this in the other PR

**Comment by maxnoe** [comment]
Superseded by #121

**Pull request closed without merging**
